File: apps/somi-home-ios/SOMIHome/Features/Today/ViewModels/TodayViewModel.swift
import Foundation
import Combine

@MainActor
final class TodayViewModel: ObservableObject {
    @Published var todayData: TodayViewResponse?
    @Published var isLoading = false
    @Published var errorMessage: String?
    @Published var pendingCount = 0

    /// Set to true (briefly) when the patient just completed a full round but still
    /// has more rounds left today. TodayView presents a congrats sheet then resets it.
    @Published var showCongratsModal = false

    private var cancellable: AnyCancellable?

    init() {
        cancellable = CompletionSyncManager.shared.$pendingCount
            .receive(on: DispatchQueue.main)
            .assign(to: \.pendingCount, on: self)
    }

    // MARK: - Computed helpers

    /// How many full rounds have been completed today (all exercises done for that occurrence).
    var completedRoundsToday: Int {
        guard let data = todayData, !data.assignments.isEmpty else { return 0 }
        var count = 0
        for round in 1...data.timesPerDay {
            let allDone = data.assignments.allSatisfy { assignment in
                assignment.completions.first { $0.occurrence == round }?.completed == true
            }
            if allDone { count += 1 }
        }
        return count
    }

    /// True when every occurrence of every exercise is marked complete.
    var isAllDoneForDay: Bool {
        guard let data = todayData, !data.assignments.isEmpty else { return false }
        return completedRoundsToday == data.timesPerDay
    }

    /// The active round: the first incomplete occurrence, or timesPerDay when all rounds are done.
    var currentOccurrence: Int {
        guard let data = todayData, !data.assignments.isEmpty else { return 1 }
        for round in 1...data.timesPerDay {
            let allDone = data.assignments.allSatisfy { assignment in
                assignment.completions.first { $0.occurrence == round }?.completed == true
            }
            if !allDone { return round }
        }
        return data.timesPerDay
    }

    // MARK: - Load

    func loadToday() async {
        isLoading = true
        errorMessage = nil

        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let dateLocal = formatter.string(from: Date())

        do {
            let response: TodayViewResponse = try await APIClient.shared.fetch(
                Endpoint.getToday(dateLocal: dateLocal)
            )
            todayData = response
        } catch APIError.serverError(404) {
            // No published treatment plan yet — show friendly empty state
            todayData = nil
        } catch APIError.networkUnavailable {
            errorMessage = "No internet connection."
        } catch {
            errorMessage = "Failed to load exercises."
        }

        isLoading = false
        CompletionSyncManager.shared.updatePendingCount()
    }

    // MARK: - Complete / Incomplete

    func markComplete(assignmentKey: String, exerciseVersionId: String, occurrence: Int) async {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let dateLocal = formatter.string(from: Date())

        let prevRounds = completedRoundsToday

        // Optimistic update
        setCompletion(assignmentKey: assignmentKey, occurrence: occurrence, completed: true)

        // Detect round completion
        let newRounds = completedRoundsToday
        if newRounds > prevRounds, let data = todayData {
            if newRounds < data.timesPerDay {
                // Completed a round, but more to go — show congrats
                showCongratsModal = true
            }
            // If newRounds == timesPerDay the TodayView shows the all-done state automatically
        }

        let endpoint = Endpoint.postCompletion(
            dateLocal: dateLocal,
            occurrence: occurrence,
            exerciseVersionId: exerciseVersionId,
            source: "mobile_ios",
            idempotencyKey: UUID().uuidString
        )

        do {
            try await APIClient.shared.fetchVoid(endpoint)
        } catch APIError.networkUnavailable {
            // Offline: keep optimistic state, enqueue for later sync
            LocalCompletionStore.shared.enqueue(
                dateLocal: dateLocal,
                occurrence: occurrence,
                exerciseVersionId: exerciseVersionId,
                source: "mobile_ios"
            )
            CompletionSyncManager.shared.updatePendingCount()
        } catch {
            // Revert optimistic update
            setCompletion(assignmentKey: assignmentKey, occurrence: occurrence, completed: false)
            errorMessage = "Failed to record completion."
        }
    }

    func markIncomplete(assignmentKey: String, exerciseVersionId: String, occurrence: Int) async {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let dateLocal = formatter.string(from: Date())

        // Optimistic update
        setCompletion(assignmentKey: assignmentKey, occurrence: occurrence, completed: false)

        let endpoint = Endpoint.deleteCompletion(
            dateLocal: dateLocal,
            occurrence: occurrence,
            exerciseVersionId: exerciseVersionId
        )

        do {
            try await APIClient.shared.fetchVoid(endpoint)
        } catch APIError.networkUnavailable {
            // Revert — unmark requires connectivity
            setCompletion(assignmentKey: assignmentKey, occurrence: occurrence, completed: true)
            errorMessage = "Cannot undo completion while offline."
        } catch {
            // Revert on failure
            setCompletion(assignmentKey: assignmentKey, occurrence: occurrence, completed: true)
        }
    }

    func refresh() async {
        await CompletionSyncManager.shared.syncPendingCompletions()
        await loadToday()
    }

    // MARK: - Optimistic helpers

    private func setCompletion(assignmentKey: String, occurrence: Int, completed: Bool) {
        guard var data = todayData else { return }
        for ai in data.assignments.indices
            where data.assignments[ai].assignmentKey == assignmentKey {
            for ci in data.assignments[ai].completions.indices
                where data.assignments[ai].completions[ci].occurrence == occurrence {
                data.assignments[ai].completions[ci] = CompletionEntry(
                    occurrence: occurrence,
                    completed: completed,
                    completedAt: completed ? ISO8601DateFormatter().string(from: Date()) : nil
                )
            }
        }
        todayData = data
    }
}

File: apps/somi-home-android/app/src/main/java/com/somi/home/features/today/TodayViewModel.kt
package com.somi.home.features.today

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.work.WorkManager
import com.somi.home.core.connectivity.ConnectivityObserver
import com.somi.home.core.database.PendingCompletionDao
import com.somi.home.core.database.PendingCompletionEntity
import com.somi.home.core.models.CompletionRequest
import com.somi.home.core.models.TodayAssignment
import com.somi.home.core.models.TodayViewResponse
import com.somi.home.core.models.UncompletionRequest
import com.somi.home.core.network.ApiService
import com.somi.home.core.sync.CompletionSyncWorker
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.util.UUID
import javax.inject.Inject

sealed class TodayUiState {
    data object Loading : TodayUiState()
    data class Success(
        val data: TodayViewResponse,
        val errorMessage: String? = null,
        val showCongratsModal: Boolean = false,
        val completedRoundsForModal: Int = 0
    ) : TodayUiState()
    data class Error(val message: String) : TodayUiState()
}

@HiltViewModel
class TodayViewModel @Inject constructor(
    private val apiService: ApiService,
    private val dao: PendingCompletionDao,
    private val connectivityObserver: ConnectivityObserver,
    private val workManager: WorkManager
) : ViewModel() {

    private val _uiState = MutableStateFlow<TodayUiState>(TodayUiState.Loading)
    val uiState: StateFlow<TodayUiState> = _uiState.asStateFlow()

    val pendingCount: StateFlow<Int> = dao.getPendingCount()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), 0)

    init {
        loadToday()
    }

    fun loadToday() {
        viewModelScope.launch(Dispatchers.IO) {
            _uiState.value = TodayUiState.Loading
            try {
                val dateLocal = LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)
                val response = apiService.getToday(dateLocal)
                _uiState.value = TodayUiState.Success(response)
            } catch (e: Exception) {
                _uiState.value = TodayUiState.Error(
                    e.message ?: "Failed to load today's exercises"
                )
            }
        }
    }

    fun markComplete(
        sessionKey: String,
        assignmentKey: String,
        exerciseVersionId: String,
        occurrence: Int
    ) {
        val currentState = _uiState.value
        if (currentState !is TodayUiState.Success) return

        val prevRounds = completedRoundsCount(currentState.data)

        // Optimistic update
        val updatedData = applyOptimisticCompletion(
            currentState.data, sessionKey, assignmentKey, occurrence
        )

        val newRounds = completedRoundsCount(updatedData)
        // Show congrats modal when a round completes but more rounds remain
        val showModal = newRounds > prevRounds && newRounds < updatedData.timesPerDay

        _uiState.value = TodayUiState.Success(
            data = updatedData,
            showCongratsModal = showModal,
            completedRoundsForModal = if (showModal) newRounds else 0
        )

        val dateLocal = LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)
        val idempotencyKey = UUID.randomUUID().toString()

        viewModelScope.launch(Dispatchers.IO) {
            if (connectivityObserver.isCurrentlyOnline()) {
                try {
                    apiService.postCompletion(
                        idempotencyKey = idempotencyKey,
                        body = CompletionRequest(
                            dateLocal = dateLocal,
                            occurrence = occurrence,
                            exerciseVersionId = exerciseVersionId,
                            source = "android"
                        )
                    )
                } catch (e: Exception) {
                    _uiState.value = TodayUiState.Success(
                        currentState.data,
                        errorMessage = "Failed to save completion. Please try again."
                    )
                }
            } else {
                dao.insert(
                    PendingCompletionEntity(
                        id = UUID.randomUUID().toString(),
                        dateLocal = dateLocal,
                        occurrence = occurrence,
                        exerciseVersionId = exerciseVersionId,
                        idempotencyKey = idempotencyKey,
                        source = "android",
                        createdAt = System.currentTimeMillis()
                    )
                )
                workManager.enqueue(CompletionSyncWorker.buildRequest())
            }
        }
    }

    fun markIncomplete(
        sessionKey: String,
        assignmentKey: String,
        exerciseVersionId: String,
        occurrence: Int
    ) {
        val currentState = _uiState.value
        if (currentState !is TodayUiState.Success) return

        // Optimistic update
        val updatedData = applyOptimisticUncomplete(
            currentState.data, sessionKey, assignmentKey, occurrence
        )
        _uiState.value = TodayUiState.Success(updatedData)

        val dateLocal = LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)

        viewModelScope.launch(Dispatchers.IO) {
            if (connectivityObserver.isCurrentlyOnline()) {
                try {
                    apiService.deleteCompletion(
                        UncompletionRequest(
                            dateLocal = dateLocal,
                            occurrence = occurrence,
                            exerciseVersionId = exerciseVersionId
                        )
                    )
                } catch (e: Exception) {
                    _uiState.value = TodayUiState.Success(
                        currentState.data,
                        errorMessage = "Failed to undo completion."
                    )
                }
            }
        }
    }

    fun dismissCongratsModal() {
        val state = _uiState.value as? TodayUiState.Success ?: return
        _uiState.value = state.copy(showCongratsModal = false, completedRoundsForModal = 0)
    }

    fun clearErrorMessage() {
        val currentState = _uiState.value
        if (currentState is TodayUiState.Success) {
            _uiState.value = currentState.copy(errorMessage = null)
        }
    }

    /**
     * The active session round: the first occurrence where not all exercises are done.
     * This is the occurrence that all exercises should be working toward.
     */
    fun currentOccurrence(data: TodayViewResponse): Int {
        for (round in 1..data.timesPerDay) {
            val allDone = data.assignments.all { a ->
                a.completions.any { it.occurrence == round && it.completed }
            }
            if (!allDone) return round
        }
        return data.timesPerDay
    }

    /**
     * True when every occurrence of every exercise is marked complete.
     */
    fun isAllDoneForDay(data: TodayViewResponse): Boolean {
        if (data.assignments.isEmpty()) return false
        return completedRoundsCount(data) >= data.timesPerDay
    }

    private fun completedRoundsCount(data: TodayViewResponse): Int {
        var count = 0
        for (round in 1..data.timesPerDay) {
            val allDone = data.assignments.all { a ->
                a.completions.any { it.occurrence == round && it.completed }
            }
            if (allDone) count++
        }
        return count
    }

    private fun applyOptimisticCompletion(
        data: TodayViewResponse,
        sessionKey: String,
        assignmentKey: String,
        occurrence: Int
    ): TodayViewResponse {
        if (data.sessionKey != sessionKey) return data
        val updatedAssignments = data.assignments.map { assignment ->
            if (assignment.assignmentKey == assignmentKey) {
                assignment.copy(
                    completions = assignment.completions.map { entry ->
                        if (entry.occurrence == occurrence) entry.copy(
                            completed = true,
                            completedAt = java.time.Instant.now().toString()
                        )
                        else entry
                    }
                )
            } else {
                assignment
            }
        }
        return data.copy(assignments = updatedAssignments)
    }

    private fun applyOptimisticUncomplete(
        data: TodayViewResponse,
        sessionKey: String,
        assignmentKey: String,
        occurrence: Int
    ): TodayViewResponse {
        if (data.sessionKey != sessionKey) return data
        val updatedAssignments = data.assignments.map { assignment ->
            if (assignment.assignmentKey == assignmentKey) {
                assignment.copy(
                    completions = assignment.completions.map { entry ->
                        if (entry.occurrence == occurrence) entry.copy(completed = false, completedAt = null)
                        else entry
                    }
                )
            } else {
                assignment
            }
        }
        return data.copy(assignments = updatedAssignments)
    }

    /**
     * Find an assignment by key in the current session.
     */
    fun findAssignment(assignmentKey: String): TodayAssignment? {
        val currentState = _uiState.value
        if (currentState !is TodayUiState.Success) return null
        return currentState.data.assignments.find { it.assignmentKey == assignmentKey }
    }
}
